datashuttle/utils_mod/canonical_configs.py

```diff
@@ -0,0 +1,188 @@
+format of the configs class. This is clearly defined
+as configs can be provided from file or input dynamically
+and so careful checks must be done.
+
+If adding a new config, first add the key to
+get_canonical_config_dict( and type to
+get_canonical_config_required_types()
+"""
+
+from pathlib import Path
+from typing import Union, get_args
+
+from datashuttle.utils_mod import canonical_configs, utils
+
+
+def get_canonical_config_dict():
+    """
+    The only permitted keys in the
+    DataShuttle config.
+    """
+    config_dict = {
+        "local_path": None,
+        "remote_path": None,
+        "connection_method": None,
+        "remote_host_id": None,
+        "remote_host_username": None,
+        "use_ephys": None,
+        "use_behav": None,
+        "use_imaging": None,
+        "use_histology": None,
+    }
+    return config_dict
+
+
+def get_experiment_types():
+    return ["use_ephys", "use_behav", "use_imaging", "use_histology"]
+
+
+def get_canonical_config_required_types():
+    """
+    The only permitted types for DataShuttle
+    config values.
+    """
+    required_types = {
+        "local_path": Union[str, Path],
+        "remote_path": Union[str, Path],
+        "connection_method": str,
+        "remote_host_id": Union[str, None],
+        "remote_host_username": Union[str, None],
+        "use_ephys": bool,
+        "use_behav": bool,
+        "use_imaging": bool,
+        "use_histology": bool,
+    }
+
+    assert (
+        required_types.keys() == get_canonical_config_dict().keys()
+    ), "update get_canonical_config_required_types with required types."
+
+    return required_types
+
+
+def check_dict_values_and_inform_user(config_dict):
+    """
+    Central function for performing checks on a
+    DataShuttle Configs UserDict class. This should
+    be run after any change to the configs (e.g.
+    make_config_file, update_config, supply_config_file).
+
+    This will raise assert if condition is not met.
+    """
+    canonical_dict = get_canonical_config_dict()
+
+    for key in canonical_dict.keys():
+        if key not in config_dict.keys():
+            utils.raise_error(
+                f"Loading Failed. The key {key} was not "
+                f"found in the supplied config. "
+                f"Config file was not updated."
+            )
+
+    for key in config_dict.keys():
+        if key not in canonical_dict.keys():
+            utils.raise_error(
+                f"The supplied config contains an invalid key: {key}. "
+                f"Config file was not updated."
+            )
+
+    check_config_types(config_dict)
+
+    if list(config_dict.keys()) != list(canonical_dict.keys()):
+        utils.raise_error(
+            f"New config keys are in the wrong order. The"
+            f" order should be: {canonical_dict.keys()}"
+        )
+
+    if config_dict["connection_method"] not in ["ssh", "local_filesystem"]:
+        utils.raise_error("connection method must be ssh or local")
```

Comment:
Discrepancy between the if statement and the error message: is it "local" or "local_filesystem"?